### ISSUE TITLE
Display Ads: Target by Tag on Posts with matching Tag(s)

### ISSUE THREAD

**Issue by mirie** [forem/forem#18519]
## Problem Statement
As an Admin
I want to specify which tags a Display Ad should appear on Posts with matching tags
So that I can increase conversion on my Display Ads

### Requirements:

**Timebox 1/2 day to investigating whether or not the tag selector component can be used. Fallback = a text field, comma separated tags, max 10 items, with the assumption that the admin has to enter the tag as an exact match (get spelling correct/entering an existing tag). IF we need to go with text field, then we'll need to check on how we want to handle validation of tags entered [i.e., what to do if tag doesn't exist]** 

- When creating or updating a Display Ad, there should be a field to specify which tags the Display Ad will show up on a Post in the `post_comments` area.
  - Appears under `Placement Area` field
  - Field name: `Targeted Tag(s)`
  - Optional field
  - Field should **function** the same way as this tag selection component except limit **10** tags:
![Image](https://user-images.githubusercontent.com/896728/193174638-ff80cb6d-9b80-443a-a1c3-ae24bfbd3f96.png)
  **- If this ^ is too complex to do due to code re-usability problems, a text field is an acceptable MVP** 
   - Only applies when Placement Area is `Below the comment section` -- Ideal UX, show/hide field when `Below the comment section` is selected. 
   - There should be form validation feedback (error message) to let the Admin know: "You can only target Tags for the following Placement Area(s): 'Below the comment section'. Please adjust the Placement Area or remove Tags from Targeted Tag(s)"
   - There can be multiple display ads that target the same tag: we want to round robin the display of this set
  - Other notes: 
    - No need to worry about doing anything special regarding: Un/Supported Tags or Require Approval tags. We will allow an Admin to select any Tag
    - If no Tags are set, behaves exactly the same as today without this feature (basically no additional targeting/limiting of where the Display Ad appears)
    - No difference in behavior for anonymous vs logged in users for this particular field -- that is managed by the `Display to user group` field
    - In the future we will make other placement areas taggable, but for now it is just this one location, so we should avoid hard-coding when possible (example, form validation error message, etc)
    - In the future, we may want to show Display Ads (with Tags set) on the home page and target all followers of the Tag(s)
- Adjust display query logic so that display_ads tagged with certain Tags only show up on Posts that have at least one of those Tags.

#### Other Technical Notes from Ben

- Unless we have a great reason not to, we should re-use existing `ActsAsTaggableOn` functionality, and treat that as a core building block.
- Ensure tag list is maximally optimized to replicate what we have done on `articles`
    - Calling this out because we have really optimized the queries on articles, and we should copy all of that over to `DisplayAds`. This will involve adding a new cache column and index in addition to the drop-in-place `ActsAsTaggableOn` functionality.

## Definition of Done

- [x] An Admin can create a Display Ad and select up to 10 Tags when the placement area is `Below the comment section`
- [x] An Admin can update a Display Ad and select up to 10 Tags when the placement area is `Below the comment section`
- [ ] There is form validation and feedback when creating/updating an existing Display Ad when an unsupported placement area is selected and Tags are selected
- [ ] For a Display Ad with Tags set, the Display Ad should appear on Posts with at least one matching Tag
- [ ] For a Display Ad with no Tags set, there should be no change in how the Display Ad is show to the end user from how it operates today

## Product Point Person / Stakeholder
@pkfrank / @noah-brinker 

## Release plan

- QA is going to be crucial prior to release: Coordinate with Peter on QA plan/test data/etc.
- Let Noah is aware that this functionality is getting close to release and when it's deployed to Prod
- Demo internally at an All Hands

**Comment by Ridhwana** [comment]
@pkfrank/ @benhalpern I'm reading this excerpt 
```
In the future, we may want to show Display Ads (with Tags set) on the home page and target all followers of the Tag(s) 
```

I'm not sure I understand what we mean by this, I'm wondering where on the home page would we possibly want to do this in the future. Would we show the display ads based on the tags shown on the homepage right-hand sidebar?

I'm asking this just so that I can ensure that the architecture is flexible enough to account for this and some other use cases in the future.

**Comment by benhalpern** [comment]
@Ridhwana I think that is partially correct, but the last point is different than the intended logic.

(Though this is a _great_ question because I don't think this was made clear).

- If a Display Ad has no tags, it is not filtered at all on tags, and all posts can display the Ad.
- If a Display Ad _has_ tags, then any post with any of those tags will display the Ad.


> but will not show Display Ads that contain tags that do not match the post.

Display Ads tagged should effectively be `OR` logic, not `AND` logic... Meaning the more tags I put in my Display Ad, the more I _broaden_ the number of posts which will likely qualify.

In the future we could hypothetically offer the logic to declare whether we want to target to "any of these tags present", or to "all of these tags present", but for now we are only serving "any of these tags present". If I tag an ad `javascript` and `ruby`, then it will display to all posts tagged javascript or ruby, it does not require both javascript and ruby to be present.

***

The broadest possible query is no tags at all. But once I add any tags, each additional tag actually re-broaden's the query. Hypothetically "no tags" is the same as "all tags". Something tagged `react` is _more narrow_ than something tagged `react` _and_ `javascript`.

I hope that makes sense. Let me know if not and we'll triple check that we're on the same page.

**Comment by Ridhwana** [comment]
Thanks so much Ben. 

A quick clarification: 

I have post that contains the following tags; #career #linux #java. Based on the above, we will show any display ads on that post that have #career or #linux or #java as a targeted tag. 

My clarifying question is, on that  same post will we also show the display ads that have no targeted tags associated with them (i.e. the admin has chosen not to add any tags for that display ad)?

**Comment by pkfrank** [comment]
Hi @Ridhwana , the steps you've outlined here sound great.  Nothing else requires coordination at this stage from my perspective.  Once we get it up and running and can QA on a Canary instance, we can think through any _final/final_ steps that we may need to think through before enabling it on DEV.

**Comment by Ridhwana** [comment]
Some notes about follow up items on this issue: 


1. Consider how DisplayAd tags might impact display_ad_event tracking? We aren't currently tracking anything related to tags on the display ad event, but it's potentially relevant to ad success rate to know which tags performed better. 
2. Refactor the `for_display` method - I've already added [this](https://github.com/forem/forem/issues/18696) as a separate issue to tackle.
3. Extract out and clean up the article factory - @maestromac you showed interest in this, is this potentially an issue you add and work on separately?
4. [Chat more](https://forem-team.slack.com/archives/CAA200SH5/p1667381470154399) about the Basic Markdown editor and what are our plans around it. I pulled some stats on editor usage and it seems that 150,994 users use the v1 editor and 801,218 users use v2.

**Comment by Ridhwana** [comment]
This [PR](https://github.com/forem/forem/pull/18710) to remove the feature flag is ready for when we’ve enabled the feature flag on DEV and tried out our first tagged ad

**Comment by mirie** [comment]
@Ridhwana does this mean that this issue can be moved out of QA to Done?

**Comment by Ridhwana** [comment]
It is done in terms of the work - however I was going to move it to done only once I remove the Feature Flags. I'll default to your judgement here.

**Comment by mirie** [comment]
The `display_ad_tags` feature flag has been enabled and there are published display ads being used successfully on dev.to:

![Image](https://user-images.githubusercontent.com/896728/202013595-1f61c3f7-3535-4b51-b001-238613b89408.png)

![Image](https://user-images.githubusercontent.com/896728/202013596-9226b7fb-c567-4f81-81bc-36fd9ff0613d.png)

https://forem.team/ben/comment/9gf

**Comment by Ridhwana** [comment]
Thanks @mirie, I've moved the PR to ready for review and will merge it as soon as possible.
